diesel/src/sqlite/connection/functions.rs

```diff
@@ -99,21 +98,69 @@ where
     })
 }
 
+#[derive(Clone)]
 struct FunctionRow<'a> {
     args: &'a [*mut ffi::sqlite3_value],
 }
 
-impl<'a> Row<Sqlite> for FunctionRow<'a> {
-    fn take(&mut self) -> Option<&SqliteValue> {
-        self.args.split_first().and_then(|(&first, rest)| {
-            self.args = rest;
-            unsafe { SqliteValue::new(first) }
+impl<'a> FunctionRow<'a> {
+    fn new(args: &'a [*mut ffi::sqlite3_value]) -> Self {
+        Self { args }
+    }
+}
+
+impl<'a> Row<'a, Sqlite> for FunctionRow<'a> {
+    type Field = FunctionArgument<'a>;
+    type InnerPartialRow = Self;
+
+    fn field_count(&self) -> usize {
+        self.args.len()
+    }
+
+    fn get<I>(&self, idx: I) -> Option<Self::Field>
+    where
+        Self: crate::row::RowIndex<I>,
+    {
+        let idx = self.idx(idx)?;
+
+        self.args.get(idx).map(|arg| FunctionArgument {
+            arg: *arg,
+            p: PhantomData,
         })
     }
 
-    fn next_is_null(&self, count: usize) -> bool {
-        self.args[..count]
-            .iter()
-            .all(|&p| unsafe { SqliteValue::new(p) }.is_none())
+    fn partial_row(&self, range: std::ops::Range<usize>) -> PartialRow<Self::InnerPartialRow> {
+        PartialRow::new(self, range)
+    }
+}
+
+impl<'a> RowIndex<usize> for FunctionRow<'a> {
+    fn idx(&self, idx: usize) -> Option<usize> {
+        Some(idx)
```

Comment:
I don't think we need a bound check here. We store a slice of arguments, accessing an argument in the slice will bound-check anyway, so no need to do a double bound check here.

Reply:
I doubt we can measure any significant performance improvement there, especially considering (i've run tests) that the double check gets optimized-out provided `RowIndex::idx` gets inlined (which we could enforce, but my guess would be the optimizer naturally does it anyway).
Overall I don't feel like it's worth having the added complexity that `RowIndex` does not guarantee the row would actually exist, though given the signature it looks that's what it would do. It would also make it consistent with the other implementations.